sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java

```diff
@@ -324,24 +315,27 @@ public void putDecimal(int rowId, Decimal value, int precision) {
   @Override
   public UTF8String getUTF8String(int rowId) {
     if (dictionary == null) {
-      ColumnarArray a = getByteArray(rowId);
-      return UTF8String.fromBytes(a.byteArray, a.byteArrayOffset, a.length);
+      return childColumns[0].getUTF8String0(getArrayOffset(rowId), getArrayLength(rowId));
     } else {
       byte[] bytes = dictionary.decodeToBinary(dictionaryIds.getDictId(rowId));
       return UTF8String.fromBytes(bytes);
```

Comment:
hmm, but looks `decodeToBinary` will copy byte data?

Reply:
That seems orthogonal to this issue. It would be nice if we could avoid the copy though. That would require some work on the dictionary code path.